File: ts/src/middleware.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { globMatch } from "./middleware.js";

class MockRes {
  statusCode = 200;
  headers: Record<string, string | string[] | number> = {};
  bodyChunks: string[] = [];
  headersSent = false;
  setHeader(name: string, value: string | string[] | number) {
    this.headers[name.toLowerCase()] = value;
  }
  getHeader(name: string) {
    return this.headers[name.toLowerCase()];
  }
  end(chunk?: string) {
    if (chunk) this.bodyChunks.push(chunk);
    this.headersSent = true;
  }
  writeHead(s: number) { this.statusCode = s; }
  get body() { return this.bodyChunks.join(""); }
  get setCookie(): string | string[] | undefined {
    const v = this.headers["set-cookie"];
    if (typeof v === "string" || Array.isArray(v)) return v;
    return undefined;
  }
}

function mkReq(opts: { url: string; method?: string; headers?: Record<string, string>; body?: unknown }) {
  return {
    url: opts.url,
    method: opts.method ?? "GET",
    headers: opts.headers ?? {},
    body: opts.body,
  };
}

/** Fetch that auto-services platform-token and forwards the rest to the handler. */
function autoFetch(handler: (url: string, init: RequestInit | undefined) => Promise<Response> | Response): typeof fetch {
  return (async (input: RequestInfo | URL, init?: RequestInit) => {
    const url = typeof input === "string" ? input : input.toString();
    if (url.endsWith("/auth/platform-token")) {
      return new Response(JSON.stringify({ platform_token: "pt_x", expires_in: 300 }), {
        status: 200, headers: { "content-type": "application/json" },
      });
    }
    return handler(url, init);
  }) as typeof fetch;
}

function makeRealm(handler: (url: string, init: RequestInit | undefined) => Promise<Response> | Response) {
  return createRealm({
    realmId: "r1",
    apiKey: "rk_live_x",
    baseUrl: "https://auth.test",
    origin: "https://app.test",
    fetch: autoFetch(handler),
  });
}

test("globMatch: handles single-segment and double-segment", () => {
  assert.equal(globMatch("/health", "/health"), true);
  assert.equal(globMatch("/public/*", "/public/foo"), true);
  assert.equal(globMatch("/public/*", "/public/foo/bar"), false);
  assert.equal(globMatch("/public/**", "/public/foo/bar"), true);
  assert.equal(globMatch("/webhooks/*", "/api/x"), false);
});

test("middleware: exempt path passes through to next", async () => {
  let nextCalled = false;
  const realm = makeRealm(() => new Response("{}", { status: 200 }));
  const mw = realm.middleware({ exemptPaths: ["/health"] });
  const req = mkReq({ url: "/health" });
  const res = new MockRes();
  await new Promise<void>((resolve, reject) => {
    mw(req, res as unknown as never, (err?: unknown) => {
      if (err) return reject(err);
      nextCalled = true;
      resolve();
    });
  });
  assert.equal(nextCalled, true);
  assert.equal(res.headersSent, false);
});

test("middleware: login route handles request and sets cookie", async () => {
  const realm = makeRealm((url) => {
    assert.match(url, /\/auth\/login$/);
    return new Response(JSON.stringify({
      access_token: "at",
      refresh_token: "rt-cookie",
      expires_in: 900,
      user: { id: "u1" },
      tenants: [{ id: "t1", role: "owner" }],
    }), { status: 200, headers: { "content-type": "application/json" } });
  });
  const mw = realm.middleware();
  const req = mkReq({
    url: "/login",
    method: "POST",
    headers: { "content-type": "application/json" },
    body: { method: "firebase", provider_token: "id_x" },
  });
  const res = new MockRes();
  let nextErr: unknown;
  await new Promise<void>((resolve) => {
    mw(req, res as unknown as never, (e?: unknown) => { nextErr = e; resolve(); });
    setTimeout(() => resolve(), 50);
  });
  assert.equal(nextErr, undefined);
  assert.equal(res.statusCode, 200);
  const parsed = JSON.parse(res.body);
  assert.equal(parsed.access_token, "at");
  assert.equal(parsed.refresh_token, undefined, "cookie mode must not echo refresh_token in body");
  const cookie = res.setCookie;
  assert.ok(cookie, "expected set-cookie");
  const cookieStr = Array.isArray(cookie) ? cookie.join(";") : String(cookie);
  assert.match(cookieStr, /realmid_refresh=rt-cookie/);
  assert.match(cookieStr, /HttpOnly/);
  assert.match(cookieStr, /SameSite=Lax/);
});

test("middleware: tokenDelivery=body returns refresh_token in JSON, no cookie", async () => {
  const realm = makeRealm(() => new Response(JSON.stringify({
    access_token: "at",
    refresh_token: "rt-body",
    expires_in: 900,
    user: { id: "u1" },
    tenants: [],
  }), { status: 200, headers: { "content-type": "application/json" } }));
  const mw = realm.middleware({ tokenDelivery: "body" });
  const req = mkReq({
    url: "/login", method: "POST",
    headers: { "content-type": "application/json" },
    body: { method: "firebase", provider_token: "id_x" },
  });
  const res = new MockRes();
  await new Promise<void>((resolve) => {
    mw(req, res as unknown as never, () => resolve());
    setTimeout(resolve, 50);
  });
  assert.equal(res.statusCode, 200);
  const parsed = JSON.parse(res.body);
  assert.equal(parsed.access_token, "at");
  assert.equal(parsed.refresh_token, "rt-body");
  assert.equal(res.setCookie, undefined, "body mode must not set a refresh cookie");
});

test("middleware: unauthenticated request gets 401 envelope", async () => {
  const realm = makeRealm(() => new Response("{}", { status: 200 }));
  const mw = realm.middleware();
  const req = mkReq({ url: "/me" });
  const res = new MockRes();
  await new Promise<void>((resolve) => {
    mw(req, res as unknown as never, () => resolve());
    setTimeout(resolve, 50);
  });
  assert.equal(res.statusCode, 401);
  const parsed = JSON.parse(res.body);
  assert.equal(parsed.error.code, "unauthorized");
});

// ---- Helpers for signed-token tests ----

async function mintSignedToken(opts: {
  realmId: string; baseUrl: string; aud: string; extraClaims?: Record<string, unknown>;
}): Promise<{ token: string; publicJwk: object }> {
  const kp = await globalThis.crypto.subtle.generateKey(
    { name: "RSASSA-PKCS1-v1_5", modulusLength: 2048, publicExponent: new Uint8Array([1, 0, 1]), hash: "SHA-256" },
    true, ["sign", "verify"],
  );
  const jwk = await globalThis.crypto.subtle.exportKey("jwk", kp.publicKey);
  const publicJwk = { kty: jwk.kty!, n: jwk.n!, e: jwk.e!, kid: "k1", alg: "RS256", use: "sig" };
  const enc = (o: unknown) => {
    const bytes = new TextEncoder().encode(JSON.stringify(o));
    let bin = "";
    for (const b of bytes) bin += String.fromCharCode(b);
    return btoa(bin).replace(/=+$/, "").replace(/\+/g, "-").replace(/\//g, "_");
  };
  const header = enc({ alg: "RS256", typ: "JWT", kid: "k1" });
  const now = Math.floor(Date.now() / 1000);
  const payload = enc({
    iss: `${opts.baseUrl}/${opts.realmId}`,
    sub: "u1", aud: opts.aud, exp: now + 900, iat: now,
    tenant_id: "t1", role: "owner",
    ...opts.extraClaims,
  });
  const sig = await globalThis.crypto.subtle.sign(
    { name: "RSASSA-PKCS1-v1_5" }, kp.privateKey, new TextEncoder().encode(`${header}.${payload}`),
  );
  let bin = "";
  for (const b of new Uint8Array(sig)) bin += String.fromCharCode(b);
  const sigStr = btoa(bin).replace(/=+$/, "").replace(/\+/g, "-").replace(/\//g, "_");
  return { token: `${header}.${payload}.${sigStr}`, publicJwk };
}

test("middleware: valid bearer attaches claims to req.realmid", async () => {
  const realmId = "r1";
  const baseUrl = "https://auth.test";
  const aud = "example.com";
  const { token, publicJwk } = await mintSignedToken({ realmId, baseUrl, aud });

  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const url = typeof input === "string" ? input : input.toString();
    if (url.endsWith("/auth/platform-token")) {
      return new Response(JSON.stringify({ platform_token: "pt_x", expires_in: 300 }), { status: 200 });
    }
    if (url.endsWith(`/${realmId}/.well-known/jwks.json`)) {
      return new Response(JSON.stringify({ keys: [publicJwk] }), { status: 200 });
    }
    return new Response("not found", { status: 404 });
  }) as typeof fetch;

  const realm = createRealm({ realmId, apiKey: "rk_live_x", baseUrl, fetch, audience: aud });
  const mw = realm.middleware();
  const req = mkReq({ url: "/me", headers: { authorization: `Bearer ${token}` } });
  const res = new MockRes();
  let nextCalled = false;
  await new Promise<void>((resolve) => {
    mw(req as never, res as unknown as never, () => { nextCalled = true; resolve(); });
    setTimeout(resolve, 200);
  });
  assert.equal(nextCalled, true, "expected next() to be called for valid token");
  const r = req as typeof req & { realmid?: { sub?: string; tenant_id?: string } };
  assert.equal(r.realmid?.sub, "u1");
  assert.equal(r.realmid?.tenant_id, "t1");
});

test("middleware: mfaProtectedPaths returns 412 mfa_required when token lacks MFA marker", async () => {
  const realmId = "r1";
  const baseUrl = "https://auth.test";
  const aud = "example.com";
  const { token, publicJwk } = await mintSignedToken({ realmId, baseUrl, aud });

  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const url = typeof input === "string" ? input : input.toString();
    if (url.endsWith("/auth/platform-token")) {
      return new Response(JSON.stringify({ platform_token: "pt_x", expires_in: 300 }), { status: 200 });
    }
    if (url.endsWith(`/${realmId}/.well-known/jwks.json`)) {
      return new Response(JSON.stringify({ keys: [publicJwk] }), { status: 200 });
    }
    if (url.endsWith("/auth/mfa/challenge")) {
      return new Response(JSON.stringify({ mfa_challenge_token: "ch_abc", methods: ["totp"] }), {
        status: 200, headers: { "content-type": "application/json" },
      });
    }
    return new Response("not found", { status: 404 });
  }) as typeof fetch;

  const realm = createRealm({ realmId, apiKey: "rk_live_x", baseUrl, fetch, audience: aud });
  const mw = realm.middleware({ mfaProtectedPaths: ["/admin/*"] });
  const req = mkReq({ url: "/admin/me", headers: { authorization: `Bearer ${token}` } });
  const res = new MockRes();
  await new Promise<void>((resolve) => {
    mw(req as never, res as unknown as never, () => resolve());
    setTimeout(resolve, 200);
  });
  assert.equal(res.statusCode, 412);
  const parsed = JSON.parse(res.body);
  assert.equal(parsed.error.code, "mfa_required");
  assert.equal(parsed.mfa_challenge_token, "ch_abc");
});

test("middleware: mfaProtectedPaths lets through tokens carrying amr=mfa", async () => {
  const realmId = "r1";
  const baseUrl = "https://auth.test";
  const aud = "example.com";
  const { token, publicJwk } = await mintSignedToken({
    realmId, baseUrl, aud, extraClaims: { amr: ["pwd", "mfa"] },
  });

  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const url = typeof input === "string" ? input : input.toString();
    if (url.endsWith("/auth/platform-token")) {
      return new Response(JSON.stringify({ platform_token: "pt_x", expires_in: 300 }), { status: 200 });
    }
    if (url.endsWith(`/${realmId}/.well-known/jwks.json`)) {
      return new Response(JSON.stringify({ keys: [publicJwk] }), { status: 200 });
    }
    return new Response("not found", { status: 404 });
  }) as typeof fetch;

  const realm = createRealm({ realmId, apiKey: "rk_live_x", baseUrl, fetch, audience: aud });
  const mw = realm.middleware({ mfaProtectedPaths: ["/admin/*"] });
  const req = mkReq({ url: "/admin/me", headers: { authorization: `Bearer ${token}` } });
  const res = new MockRes();
  let nextCalled = false;
  await new Promise<void>((resolve) => {
    mw(req as never, res as unknown as never, () => { nextCalled = true; resolve(); });
    setTimeout(resolve, 200);
  });
  assert.equal(nextCalled, true);
  assert.equal(res.headersSent, false);
});

File: go/middleware_test.go
package realmid

import (
	"context"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"io"
	"log/slog"
	"math/big"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func mwTestServer(t *testing.T, jwks []jwk, audience string, extra map[string]http.HandlerFunc) *httptest.Server {
	t.Helper()
	mux := http.NewServeMux()
	mux.HandleFunc("/auth/platform-token", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"platform_token": "ptok", "expires_in": 300})
	})
	mux.HandleFunc("/platforms/mine", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []any{map[string]any{"id": testRealmID, "audience": audience}},
		})
	})
	mux.HandleFunc("/"+testRealmID+"/.well-known/jwks.json", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(jwksDoc{Keys: jwks})
	})
	for p, h := range extra {
		mux.HandleFunc(p, h)
	}
	return httptest.NewServer(mux)
}

func mintTestKey(t *testing.T, kid string) (signFn, jwk) {
	t.Helper()
	priv, _ := rsa.GenerateKey(rand.Reader, 2048)
	pub := priv.PublicKey
	pj := jwk{
		Kty: "RSA", Kid: kid, Alg: "RS256", Use: "sig",
		N: base64.RawURLEncoding.EncodeToString(pub.N.Bytes()),
		E: base64.RawURLEncoding.EncodeToString(big.NewInt(int64(pub.E)).Bytes()),
	}
	sign := func(claims map[string]any) string {
		hb, _ := json.Marshal(map[string]any{"alg": "RS256", "typ": "JWT", "kid": kid})
		cb, _ := json.Marshal(claims)
		signing := base64.RawURLEncoding.EncodeToString(hb) + "." + base64.RawURLEncoding.EncodeToString(cb)
		sum := sha256.Sum256([]byte(signing))
		sig, _ := rsa.SignPKCS1v15(rand.Reader, priv, crypto.SHA256, sum[:])
		return signing + "." + base64.RawURLEncoding.EncodeToString(sig)
	}
	return sign, pj
}

func TestMiddleware_ExemptPathPasses(t *testing.T) {
	srv := mwTestServer(t, nil, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{ExemptPaths: []string{"/health"}})

	called := false
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		called = true
		w.WriteHeader(200)
	}))
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, httptest.NewRequest("GET", "/health", nil))
	if !called {
		t.Fatal("exempt path should pass through")
	}
}

func TestMiddleware_LoginRouteHandledBySDK(t *testing.T) {
	srv := mwTestServer(t, nil, testAud, map[string]http.HandlerFunc{
		"/auth/login": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"access_token": "atok", "refresh_token": "rtok", "expires_in": 900,
				"user": map[string]any{"id": "u1"}, "tenants": []any{},
			})
		},
	})
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{TokenDelivery: "body"})
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(404) }))

	body := strings.NewReader(`{"method":"firebase","provider_token":"pt"}`)
	req := httptest.NewRequest("POST", "/login", body)
	req.Header.Set("Content-Type", "application/json")
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)

	if rec.Code != 200 {
		t.Fatalf("status: %d body: %s", rec.Code, rec.Body.String())
	}
	var out map[string]any
	_ = json.Unmarshal(rec.Body.Bytes(), &out)
	if out["access_token"] != "atok" {
		t.Errorf("response: %+v", out)
	}
}

func TestMiddleware_Unauthenticated401(t *testing.T) {
	srv := mwTestServer(t, nil, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{})
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(200) }))

	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, httptest.NewRequest("GET", "/me", nil))
	if rec.Code != 401 {
		t.Errorf("status: %d", rec.Code)
	}
}

func TestMiddleware_ValidBearerAttachesClaims(t *testing.T) {
	sign, pub := mintTestKey(t, "kid-1")
	srv := mwTestServer(t, []jwk{pub}, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{})

	now := time.Now().Unix()
	tok := sign(map[string]any{
		"iss": srv.URL + "/" + testRealmID,
		"sub": "u1", "aud": testAud, "iat": now, "exp": now + 600,
	})

	var got *Claims
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		got, _ = ClaimsFrom(req.Context())
		w.WriteHeader(200)
	}))
	req := httptest.NewRequest("GET", "/me", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)

	if rec.Code != 200 {
		t.Fatalf("status: %d body: %s", rec.Code, rec.Body.String())
	}
	if got == nil || got.Subject != "u1" {
		t.Errorf("claims: %+v", got)
	}
}

func TestMiddleware_MFAProtected412(t *testing.T) {
	sign, pub := mintTestKey(t, "kid-1")
	srv := mwTestServer(t, []jwk{pub}, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{MFAProtectedPaths: []MFARule{{Path: "/admin/*"}}})

	now := time.Now().Unix()
	tok := sign(map[string]any{
		"iss": srv.URL + "/" + testRealmID,
		"sub": "u1", "aud": testAud, "iat": now, "exp": now + 600,
		// no amr/acr
	})
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(200) }))
	req := httptest.NewRequest("GET", "/admin/users", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)

	if rec.Code != http.StatusPreconditionFailed {
		t.Fatalf("status: %d body: %s", rec.Code, rec.Body.String())
	}
	var out map[string]any
	_ = json.Unmarshal(rec.Body.Bytes(), &out)
	envelope, _ := out["error"].(map[string]any)
	if envelope == nil || envelope["code"] != "mfa_required" {
		t.Errorf("envelope: %+v", out)
	}
}

func TestMiddleware_GlobMatcher(t *testing.T) {
	cases := []struct {
		pat, path string
		want      bool
	}{
		{"/health", "/health", true},
		{"/health", "/healthx", false},
		{"/public/*", "/public/x", true},
		{"/public/*", "/public/x/y", false},
		{"/public/**", "/public/x/y", true},
		{"/public/**", "/public", true},
	}
	for _, c := range cases {
		got := globToRegex(c.pat).MatchString(c.path)
		if got != c.want {
			t.Errorf("glob %q vs %q: got %v want %v", c.pat, c.path, got, c.want)
		}
	}
}

// TestLoggerRedactsCredentials verifies §9 — raw API keys never appear
// in log output.
func TestLoggerRedactsCredentials(t *testing.T) {
	var buf strings.Builder
	logger := newCapturingLogger(&buf)
	srv := mwTestServer(t, nil, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{
		RealmID: testRealmID, APIKey: "rk_live_supersecret", BaseURL: srv.URL,
		Logger: logger,
	})
	// Trigger a platform-token mint (logs at info).
	_, _ = r.platformToken.get(context.Background())
	out := buf.String()
	if out == "" {
		t.Fatal("no logs captured")
	}
	if strings.Contains(out, "supersecret") {
		t.Errorf("raw API key leaked into logs:\n%s", out)
	}
}

func newCapturingLogger(w io.Writer) *slogLogger {
	return slog.New(slog.NewTextHandler(w, &slog.HandlerOptions{Level: slog.LevelDebug}))
}

type slogLogger = slog.Logger

func TestMiddleware_MFAFresh_AcceptsWithinMaxAge(t *testing.T) {
	sign, pub := mintTestKey(t, "kid-1")
	srv := mwTestServer(t, []jwk{pub}, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{
		MFAProtectedPaths: []MFARule{{Path: "/admin/*", MaxAge: 15 * time.Minute}},
	})
	now := time.Now().Unix()
	tok := sign(map[string]any{
		"iss": srv.URL + "/" + testRealmID,
		"sub": "u1", "aud": testAud, "iat": now, "exp": now + 600,
		"mfa_at": now - 60, // 1 minute ago
	})
	called := false
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { called = true; w.WriteHeader(200) }))
	req := httptest.NewRequest("GET", "/admin/users", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)
	if !called {
		t.Fatalf("handler not called: status=%d body=%s", rec.Code, rec.Body.String())
	}
}

func TestMiddleware_MFAStale_Returns412StaleReason(t *testing.T) {
	sign, pub := mintTestKey(t, "kid-1")
	srv := mwTestServer(t, []jwk{pub}, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{
		MFAProtectedPaths: []MFARule{{Path: "/admin/*", MaxAge: 15 * time.Minute}},
	})
	now := time.Now().Unix()
	tok := sign(map[string]any{
		"iss": srv.URL + "/" + testRealmID,
		"sub": "u1", "aud": testAud, "iat": now, "exp": now + 600,
		"mfa_at": now - 1800, // 30 minutes ago — stale
	})
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(200) }))
	req := httptest.NewRequest("GET", "/admin/users", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)
	if rec.Code != http.StatusPreconditionFailed {
		t.Fatalf("status: %d", rec.Code)
	}
	var out map[string]any
	_ = json.Unmarshal(rec.Body.Bytes(), &out)
	if out["reason"] != "stale_mfa" {
		t.Errorf("reason: %v", out["reason"])
	}
	if out["max_age_seconds"] != float64(900) {
		t.Errorf("max_age_seconds: %v", out["max_age_seconds"])
	}
}

func TestMiddleware_RequireFresh_RejectsLegacyMarker(t *testing.T) {
	sign, pub := mintTestKey(t, "kid-1")
	srv := mwTestServer(t, []jwk{pub}, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{
		MFAProtectedPaths: []MFARule{{Path: "/billing/*", RequireFresh: true}},
	})
	now := time.Now().Unix()
	tok := sign(map[string]any{
		"iss": srv.URL + "/" + testRealmID,
		"sub": "u1", "aud": testAud, "iat": now, "exp": now + 600,
		"amr": []string{"pwd", "mfa"}, // legacy marker, no mfa_at
	})
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(200) }))
	req := httptest.NewRequest("POST", "/billing/charge", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)
	if rec.Code != http.StatusPreconditionFailed {
		t.Fatalf("status: %d", rec.Code)
	}
	var out map[string]any
	_ = json.Unmarshal(rec.Body.Bytes(), &out)
	if out["reason"] != "fresh_required" {
		t.Errorf("reason: %v", out["reason"])
	}
}

func TestMiddleware_RequireFresh_AcceptsRecentMfaAt(t *testing.T) {
	sign, pub := mintTestKey(t, "kid-1")
	srv := mwTestServer(t, []jwk{pub}, testAud, nil)
	defer srv.Close()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	mw := r.Middleware(MiddlewareOptions{
		MFAProtectedPaths: []MFARule{{Path: "/billing/*", RequireFresh: true}},
	})
	now := time.Now().Unix()
	tok := sign(map[string]any{
		"iss": srv.URL + "/" + testRealmID,
		"sub": "u1", "aud": testAud, "iat": now, "exp": now + 600,
		"mfa_at": now - 5, // 5 sec ago — within RequireFresh window
	})
	called := false
	h := mw(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) { called = true; w.WriteHeader(200) }))
	req := httptest.NewRequest("POST", "/billing/charge", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	rec := httptest.NewRecorder()
	h.ServeHTTP(rec, req)
	if !called {
		t.Fatalf("handler not called: status=%d body=%s", rec.Code, rec.Body.String())
	}
}
